packages/@lwc/template-compiler/src/parser/html.ts

```diff
@@ -45,87 +32,17 @@ export function parseHTML(source: string) {
         );
     };
 
-    const validateClosingTag = (node: parse5.AST.Default.Element) => {
-        if (!node.__location) {
-            return;
-        }
-
-        const { startTag, endTag } = node.__location;
-        const isVoidElement = VOID_ELEMENT_SET.has(node.tagName);
-        const missingClosingTag = !!startTag && !endTag;
-
-        if (!isVoidElement && missingClosingTag) {
-            parsingErrors.push(
-                generateCompilerDiagnostic(ParserDiagnostics.NO_MATCHING_CLOSING_TAGS, {
-                    messageArgs: [node.tagName],
-                    origin: {
-                        location: {
-                            line: startTag.startLine || startTag.line,
-                            column: startTag.startCol || startTag.col,
-                            start: startTag.startOffset,
-                            length: startTag.endOffset - startTag.startOffset,
-                        },
-                    },
-                })
-            );
-        }
-    };
-
     const fragment = parse5.parseFragment(source, {
         locationInfo: true,
         onParseError,
     }) as parse5.AST.Default.DocumentFragment;
 
-    if (!parsingErrors.length) {
-        traverseHTML(fragment, {
```

Comment:
The closing tag validation can be done in the middle of the AST transformation. No need to do an extra AST traversal.